File: src/spoj/dynamic_programming/mreplbrc/MREPLBRC.c
// http://www.spoj.com/problems/MREPLBRC/

#include <stdint.h>
#include <stdio.h>

#define MAX_LEN 205
#define MOD 100000L
#define FALSE 0
#define TRUE 1

char s[MAX_LEN];
int32_t s_len;

int32_t mem[MAX_LEN][MAX_LEN];

int32_t result_is_large;

int32_t coeff(int32_t i, int32_t j) {
	if(s[i] == '?' && s[j] == '?') {
		return 3;
	}

	if((s[i] == '(' && s[j] == ')')
		|| (s[i] == '(' && s[j] == '?')
		|| (s[i] == '?' && s[j] == ')')
		|| (s[i] == '[' && s[j] == ']')
		|| (s[i] == '[' && s[j] == '?')
		|| (s[i] == '?' && s[j] == ']')
		|| (s[i] == '{' && s[j] == '}')
		|| (s[i] == '{' && s[j] == '?')
		|| (s[i] == '?' && s[j] == '}')) {
	
		return 1;
	}

	return 0;
}

int32_t solve() {
	int32_t i;
	int32_t j;
	int32_t k;
	int32_t c;
	int32_t left;
	int32_t right;
	int64_t tmp_result;

	result_is_large = FALSE;

	for(i = s_len - 1; i >= 0; --i) {
		for(j = i + 1; j < s_len; ++j) {
			mem[i][j] = 0;
			for(k = i + 1; k <= j; ++k) {
				c = coeff(i, k);
				if(c > 0) {
					left = (i + 1 > k - 1) ? 1 : mem[i + 1][k - 1];
					right = (k + 1 > j) ? 1 : mem[k + 1][j];

					tmp_result = (int64_t)mem[i][j] + (int64_t)c * (int64_t)left * (int64_t)right;
					mem[i][j] = tmp_result % MOD;

					if(mem[i][j] != tmp_result) {
						result_is_large = TRUE;
					}
				}
			}
		}
	}

	return mem[0][s_len - 1];
}

int main(){

	scanf("%d", &s_len);
	scanf("%s", s);

	result_is_large = FALSE;
	int32_t result = (s_len % 2 == 0) ? solve() : 0;
	
	if(result_is_large) {
		printf("%05d\n", result);
	} else {
		printf("%d\n", result);
	}

	return 0;
}

File: src/spoj/dynamic_programming/mreplbrc/MREPLBRC.py
# http://www.spoj.com/problems/MREPLBRC/en/

class Problem:
    coefficient_1 = set([
        ('(', ')'), ('(', '?'), ('?', ')'),
        ('[', ']'), ('[', '?'), ('?', ']'),
        ('{', '}'), ('{', '?'), ('?', '}')
    ])

    coefficient_3 = ('?', '?')

    def __init__(self, s):
        self.s = s
        self.mem = {}

    def solve(self, *args):
        if not args:
            return self.solve(0, len(self.s) - 1)

        i = args[0]
        j = args[1]

        if i > j:
            return 1
        if (i, j) in self.mem:
            return self.mem[i, j]
        self.mem[i, j] = 0
        for k in xrange(i + 1, j + 1, 2):
            c = self.coeff(i, k)
            if c > 0:
                self.mem[i, j] += c * self.solve(i + 1, k - 1) * self.solve(k + 1, j)
        return self.mem[i, j]

    def coeff(self, i, j):
        brck = (self.s[i], self.s[j])
        if brck == self.coefficient_3:
            return 3
        if brck in self.coefficient_1:
            return 1
        return 0

for s in ['()', '??', '()()()', '()()?(', '()()?)', '()()??', '(?([?)]?}?', '(??)', '??(?', '?](?', '((?)???)()',
'?((?((???]})???)??())?']:
    print s
    print Problem(s).solve()
    print
